<a href="https://colab.research.google.com/github/kimjunwooo/python/blob/main/pandas%EA%B4%80%EB%A0%A8_%EC%88%99%EC%A0%9C_ipynb%EC%9D%98_%EC%82%AC%EB%B3%B8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 모두 실행해 보시고, 다음 물음에 답하시길 바랍니다.

**[숙제 1]** os 모듈은 운영체제를 제어할 수 있다. os.chdir('/content/')은 content 디렉토리로 이동한다. os.listdir()은 현재 디렉터리에 어떤 파일들이 있는지를 알려준다. os.getcwd는 현재 작업 디렉터리를 알려준다.

In [ ]:
import os

os.chdir('/content/')

print(os.getcwd()) 

print(os.listdir()) 

os.chdir('/content/sample_data/')

print(os.listdir()) 


**[숙제 2]** 이 코드는 먼저 pandas 모듈를 pd로 import하고 targetUrl은 https://raw.githubusercontent.com/gundaminpde/2022/main/gapminder_sample.ts 주소에 있는 tsv 파일을 지칭한다. df2는 csv 파일을 읽어 오는데 이때 targetUrl을 읽어오며 이때 구분자는 tab이다.

In [ ]:
### 와~ pandas

import pandas as pd


targetUrl = "https://raw.githubusercontent.com/gundaminpde/2022/main/gapminder_sample.tsv"

df2 = pd.read_csv(targetUrl, sep='\t')


**[숙제 3]** df2 데이터에서 맨 처음부터 괄호 안 숫자까지 출력. 이때 아무것도 안 넣으면 5개만 출력

In [ ]:
print(df2.head( ))

**[숙제 4]** df2 데이터의 타입의 종류가 무엇인지 알려줌

In [ ]:
print(type(df2))

**[숙제 5]** df2 데이터의 배열의 크기기를 알려줌. 예를 들어 3행 2열 데이터 set이면 (3, 2) 출력

In [ ]:
print(df2.shape)

(1704, 6)


**[숙제 6]** 데이터의 컬럼을 보여준다

In [ ]:
print( df2.columns)


**[숙제7]** 데이터의 변경할 타입을 출력한다다

In [ ]:
print(df2.dtypes)


**[숙제8]** 데이터의 정보를 확인할 수 있다. 인덱스 정보, 메모리 사용량, column 이름, 결측값이 아닌 행의 개수, 타입 정보 등등

In [ ]:
print(df2.info())

**[숙제9]** df3 데이터의 첫행부터 5번째 행까지 내용 출력력

In [ ]:

!git clone https://github.com/gundaminpde/chow.git

df3 = pd.read_csv("./chow/ThoraricSurgery.csv", sep=',') 

print(df3.head())




In [ ]:
## csv 화일의 기본 형태

## [[1행 리스트],[2행 리스트 ],[3행 리스트] ... ] 

In [ ]:
# 판다스 자료형 : 파이썬 자료형 : 설명
#  object    : string    : 문자열
#  int64     : int       : 정수
#   float64  : float     : 소숫점 가진 숫자
# datetime64 : datetime  : 파이썬 라이브러리 datetime 반환하는 자료형 (시계열할 때)


**[숙제10]** df2 데이터의 첫행부터 5번째 행까지 내용 출력
한 칸을 빈칸으로 띄우고 df2 데이터에서 country 칼럼의 데이터를 countary_df에 저장하고 그 데이터의 타입을 출력.
countary_df의 첫행부터 5번째 행까지 내용 출력
마찬가지로 countary_df의 마지막행부터 위로 5번째 행까지 출력력

In [ ]:

print(df2.head())

print(" ")

countary_df=df2["country"]

print(type(countary_df)) 

print(countary_df.head())

print(countary_df.tail())


**[숙제11]** country, continent, year columns 데이터를 subset_df에 저장
subset_df의 데이터 타입을 출력 
subset_df의 head 출력 



In [ ]:

subset_df=df2[["country", 'continent', 'year']]

print(type(subset_df)) 

print(subset_df.head())

print(subset_df.tail())


In [ ]:

print(df2.head())

**[숙제12]** df2 데이터 첫 행을 출력:


In [ ]:
print(df2.loc[0])    

**[숙제13]** df2 데이터 첫 행을 출력:


In [ ]:
print(df2.iloc[0])

**[숙제14]** df2의 데이터 shape : (1704, 6)
df2 데이터에서 마지막 인덱스 행을 출력


In [ ]:
number_rows = df2.shape[0]

print(df2.shape)

last_row_index = number_rows - 1

print(df2.loc[last_row_index])


print(" ")

## 또 다른 방법, tail(n=1)

print("또 다른 방법, tail(n=1)")

print(df2.tail(n=1))

**[숙제15]** 
0번, 20번 30번 인덱스의 행을 출력


In [ ]:
print(df2.loc[[0,20,30]]) 



**[숙제16]**
0 ~ 4번 인덱스까지 인덱싱






In [ ]:
print(df2.loc[0:4])

**[숙제17]** iloc[-1]는 실행이 가능, loc[-1]는 실생이 불가능한 이유가 뭔가요?
그 이유는 loc는 행이나 칼럼의 label을 기준으로 하고 iloc는 인덱스의 값을 기준으로 하기 때문입니다

In [ ]:
## iloc[-1]는 가능, loc[-1]는 불가능


print(df2.iloc[-1])

print(df2.loc[-1])

**[숙제18]** 
year, pop 칼럼을 기준으로 인덱싱 후 0부터 4번 인덱스 까지 추출
밑에는 칼럼 인덱스 2번 4번 -1번 즉 year, pop gdpPercap 칼럼을 추출하고 0번부터 4번 인덱스 행까지 출력



In [ ]:
subset1 = df2.loc[:,['year' ,'pop']]

print(subset1.head() )


subset2 = df2.iloc[:,[2,4,-1]]

print(subset2.head() )

**[숙제19]** 
column 인덱스 0번 2번 4번 추출하고 0번부터 4번 행까지 추출



In [ ]:
subset3 = df2.iloc[:,range(0,6,2)]

print(subset3.head() )

subset4 = df2.iloc[:, 0:6:2]

print(subset4.head() )


**[숙제20]** 행은 0, 99, 999번째 행 추출하고 칼럼은 0번 3번 5번 칼럼을 추출
행은 10번부터 12번까지 country, lifeExp, gdpPercap 열 추출출

In [ ]:
print(df2.iloc[[0,99,999],[0,3,5]])  

print(df2.loc[10:13,['country','lifeExp','gdpPercap']])

**[숙제21]** 
df2 데이터 9번째 행까지 추출
year 칼럼을 그룹으로 묶고 lifeExp 칼럼 값의 평균을 출력력

In [ ]:
### groupby()

print(df2.head(n=10))


print(df2.groupby('year')['lifeExp'].mean())

**[숙제22]** 
먼저 year 칼럼을 기준으로 groupby를 하면 1952, 1957년 .. 으로 그룹화가 되고
too_long은 그룹화 된 데이터에서 lifeExp 칼럼만 보여준다.
마지막으로 too_long.mean()은 lifeExp 칼럼 값의 평균을 보여준다다

In [ ]:
### 위의 코드를 풀어서 설명하면...

df2_groupby_year = df2.groupby('year')
print(df2_groupby_year)
##print(df2_groupby_year.head())

too_long = df2_groupby_year['lifeExp']

too_long.mean()



**[숙제23]** 
continent 칼럼을 기준으로 그룹화를 하고 country 칼럼 값들의 고유값을 출력력

이것은 매우 중요합니다.

In [ ]:
print(df2.groupby('continent')['country'].nunique())

**[숙제24]** year로 그룹화를 하고 lifeExp의 평균 값들을 출력한다
그리고 그 값들을 선형 그래프로 나타낸다
x 축은 year y축은 lifeExp의 평균 값을 나타낸다
마지막에 or은 점 그래프를 나타낸다

In [ ]:
### matplotlib inline <==jupyter notebook용 그래프그리기 __매직함수magic method__


import matplotlib.pyplot as plt

year_lifeExpectancy = df2.groupby('year')['lifeExp'].mean()

print(year_lifeExpectancy)

year_lifeExpectancy.plot()

plt.show()

plt.plot(year_lifeExpectancy,'or')

**[숙제25]** 
'와', '어이', 123은 인덱스 0, 1, 2와 각각 연결된다
'우와', '자동차', 123은 인덱스 '감탄사', '명사', '숫자'와 연결된다

In [ ]:
import pandas as pd

A_series = pd.Series(['와','어이',123 ])

print(A_series)

print('')

B_series = pd.Series(['우와','자동차',123 ], index = ["감탄사","명사", '숫자'])

print(B_series)

**[숙제26]** 
0, 1, 2 인덱스에 정보가 주어지는데 칼럼은 이름, 키, 몸무게, 인기 이다
인덱스와 칼럼에 맞게 자료가 입력된다
ex)  
     이름  키 몸무게  인기
  0 뽀로로 1.2 가벼움 많음

In [ ]:

A_D_frame = pd.DataFrame({ 
    '이름':['뽀로로', '포비', '에디' ],
    '키(m)':[1.2 , 2.1, 1.4],
    '몸무게': ['가벼움', '무거움', '보통'],
    '인기': ['많음', '보통', '보통']}
)

print(A_D_frame)


**[숙제27]** 
위에 코드와는 다르게 이번 코드에서는 인덱스를 자동부여가 아닌 뽀로로, 포비, 에디로 따로 지정했고 칼럼도 따로 지정해줬다.
따라서 위에서는 인덱스가 0, 1, 2로 부여된 반면 이번 데이터에서는 따로 작성한 인덱스로 dataframe이 만들어진다.

In [ ]:


B_D_frame = pd.DataFrame(
    data={ '키(m)':[1.2 , 2.1, 1.4],
          '몸무게': ['가벼움', '무거움', '보통'],
          '인기': ['많음', '보통', '보통']},
    index= ['뽀로로', '포비', '에디'],
    columns=["키(m)","몸무게","인기"]
)

print(B_D_frame)


**[숙제28]** 
먼저 collections 모듈에는 순서를 가진 사전형 데이터를 사용할 수 있도록 ordereddict가 있는데 이 모듈을 이용해 데이터프레임을 만든 모습이다.
따라서 이를 실행했을때 인덱스는 0,1,2 칼럼은 이름,키,몸무게,인기로 나타나며 이에 맞게 자료가 입력된다

In [ ]:


from collections import OrderedDict

C_D_frame = pd.DataFrame(OrderedDict([
    ('이름',['뽀로로', '포비', '에디' ]),
    ('키(m)',[1.2 , 2.1, 1.4]),
    ('몸무게', ['가벼움', '무거움', '보통']),
    ('인기', ['많음', '보통', '보통'])
])
)

print(C_D_frame)

**[숙제29]** 
만들어낸 dataframe에서 첫번째 행인 뽀로로 인덱스의 행을 추출한다다

In [ ]:
D_D_frame = pd.DataFrame(
    data={ '키(m)':[1.2 , 2.1, 1.4],
          '몸무게': ['가벼움', '무거움', '보통'],
          '인기': ['많음', '보통', '보통']},
    index= ['뽀로로', '포비', '에디'],
    columns=["키(m)","몸무게","인기"]
)

first_row = D_D_frame.loc['뽀로로']

print(first_row)

**[숙제30]** 
first_row의 칼럼 값들이 나온다 [키, 몸무게, 인기]

In [ ]:


print(first_row.index)

**[숙제31]** 
frist_row의 value 값들이 나온다 [1.2, 가벼움, 많음]

In [ ]:
print(first_row.values)



**[숙제32]** 
first_row의 키의 목록을 반환한다다 [키, 몸무게, 인기]

In [ ]:
print(first_row.keys())



**[숙제33]** 
키
몸무게
가벼움움

In [ ]:
print(first_row.index[0])

print(first_row.keys()[1])

print(first_row.values[1])

**[숙제34]** 
뽀로로 1.2 #데이터에서 키값들만 추출한 모습습
포비 2.1
에디 1.4
2.1 # 키 값들에서 가장 큰값
1.2 # 최소값 
1.566 # 평균
0.4725 # 분산산

In [ ]:
height=D_D_frame['키(m)']

print(height)

print(height.max())

print(height.min())

print(height.mean())

print(height.std())  # standard deviation = std

**[숙제35]** 
csv 파일을 위에서부터 10행까지 불러오기기

In [ ]:
import pandas as pd

targetUrl_sci= "https://raw.githubusercontent.com/gundaminpde/2022/main/scientists.csv"

scientists = pd.read_csv(targetUrl_sci)

scientists.head(n=10)


**[숙제36]** 
90
59.125
age 칼럼 값들을 추출하고 그 값에서의 최대, 평균 

In [ ]:
ages = scientists['Age']

print(ages.max())
print(ages.mean())



**[숙제37]** 
위 코드는 age칼럼의 value 중에서 age칼럼 평균보다 큰 값들을 추출하고
밑에 코드는 age 칼럼에 데이터를 위에서부터 내려오면서 평균보다 큰 값은 True를 작은 값은 False를 출력한다다

아주 중요해요.

In [ ]:
print(ages[ages > ages.mean()])

print(ages > ages.mean())

**[숙제38]** 
age 평균 값보다 큰 값들을 가진 row를 출력


아주 중요해요.

In [ ]:
print(scientists[ scientists['Age'] > scientists['Age'].mean() ]) ## 불린 추출

**[39]** 
age 칼럼에 있는 값들을 출력
born, died 칼럼에 있는 값들의 type 출력력

In [ ]:
print(scientists['Age'])

print(scientists['Born'].dtype) 

print(scientists['Died'].dtype)

**[숙제40]** 
scientists 데이터에서 born 칼럼에 있는 데이터를 format에 맞게 출력
밑에 코드는 strftime을 이용해서 날짜 데이터 포멧을 변경경

날짜에 관한 것입니다. 유의하세요.

In [ ]:


born_datetime = pd.to_datetime(scientists['Born'],format='%Y-%m-%d')

print(born_datetime)

changed_one=born_datetime.dt.strftime('%m/%d/%Y')

print(changed_one)



**[숙제41]**
위에서 만든 born_datetime과 여기서 만든 died_datetime을 scientists 데이터에 각각 Born_dt, Died_dt 칼럼으로 추가가

In [ ]:

died_datatime = pd.to_datetime(scientists['Died'],format='%Y-%m-%d')

print(died_datatime)

scientists['Born_dt'], scientists['Died_dt'] = (born_datetime, died_datatime)


print(scientists.head())



**[숙제42]** 
life_days 칼럼을 추가하는데 이 칼럼의 값은 died_dt에서 born_dt를 뺀 값을 추가가

In [ ]:
scientists['life_days'] =  scientists['Died_dt'] - scientists['Born_dt']

print(scientists)

**[숙제43]** 
scientists 데이터를 셔플하고 이 데이터를 슬라이싱함함

In [ ]:
print(scientists['Age'])

import random

random.seed(2)

random.shuffle(scientists['Age'])

print(scientists['Age'])

print(scientists.iloc[:,0:4])


**[숙제44]** 
0       Rosaline Franklin
1          William Gosset
2    Florence Nightingale
3             Marie Curie
4           Rachel Carson
5               John Snow
6             Alan Turing
7            Johann Gauss
Name: Name, dtype: object

In [ ]:
## 피클로 만들어 저장하고 꺼내오기.

import os

print(os.getcwd())

os.chdir('/content/')

names = scientists['Name']

names.to_pickle('../content/scientists_names.pickle')

return_names = pd.read_pickle('../content//scientists_names.pickle')

print(return_names)

**[숙제45]** 
df_cat_A는 데이터 3개를 합치는데 row방향으로 합치고
df_cat_B는 칼럼 방향으로 합침침

매우 중요해요.

In [ ]:
# Creating data frames
df_King = pd.DataFrame({'A': ['A0', 'A1', 'A2', 'A3'],
                        'B': ['B0', 'B1', 'B2', 'B3'],
                        'C': ['C0', 'C1', 'C2', 'C3'],
                        'D': ['D0', 'D1', 'D2', 'D3']},
                        index=[0, 1, 2, 3])

df_Whang = pd.DataFrame({'A': ['A4', 'A5', 'A6', 'A7'],
                        'B': ['B4', 'B5', 'B6', 'B7'],
                        'C': ['C4', 'C5', 'C6', 'C7'],
                        'D': ['D4', 'D5', 'D6', 'D7']},
                         index=[0, 1, 2, 3])

df_ZZnag = pd.DataFrame({'A': ['A8', 'A9', 'A10', 'A11'],
                        'B': ['B8', 'B9', 'B10', 'B11'],
                        'C': ['C8', 'C9', 'C10', 'C11'],
                        'D': ['D8', 'D9', 'D10', 'D11']},
                        index=[0, 1, 2, 3])

print("\nThe DataFrame number 1\n",'-'*30, sep='')
print(df_King)
print("\nThe DataFrame number 2\n",'-'*30, sep='')
print(df_Whang)
print("\nThe DataFrame number 3\n",'-'*30, sep='')
print(df_ZZnag)

### 아래가 하고 싶은 것

df_cat_A = pd.concat([df_King,df_Whang,df_ZZnag], axis=0)
print("\nAfter concatenation along row\n",'-'*30, sep='')
print(df_cat_A)

df_cat_B = pd.concat([df_King,df_Whang,df_ZZnag], axis=1)
print("\nAfter concatenation along column\n",'-'*60, sep='')
print(df_cat_B)


**[숙제46]** 
irisData의 첫행부터 5번째 까지 그리고 마지막 행부터 위로 5번째 까지 자료 출력력

In [ ]:
import seaborn as sns

targetUrl_iris="https://raw.githubusercontent.com/gundaminpde/2022/main/iris.csv"

irisData = pd.read_csv(targetUrl_iris)

print(irisData.shape)

print(irisData.head())

irisData.tail()

**[숙제47]** 
species가 seosa, versicolor, virginica로 3가지 종류가 있는데 이들의 산점도를 나타낸다다

In [ ]:
sns.set_theme()
sns.set(rc={"figure.dpi":300}) #이미지 품질 설정
sns.set(rc={"figure.figsize":(6,3)}) ## 크기설정


## Scatter plot 산점도

sns.scatterplot(x="Sepal_Length", y ="Sepal_Width", data=irisData, hue ='Species')

**[숙제48]** 
irisData의 petal_length를 히스토그램으로 출력
이때 petal_length는 y축에 위치

In [ ]:
## 히스토그램 histogram

sns.histplot( y= "Petal_Length", data = irisData)


**[숙제49]** 
irisData의 petal_length를 히스토그램으로 출력
이때 petal_length는 x축에 위치

In [ ]:
sns.histplot( x= "Petal_Length", data = irisData)

**[숙제50]** 
히스토그램을 만들고 그 위에 kdeplot을 확인할 수 있다다

In [ ]:
sns.histplot( x= "Petal_Length", data = irisData, kde = "True")

**[숙제51]** 
3종의 데이터를 하나의 히스토그램으로 파악할 수 있다다


In [ ]:
sns.histplot( x= "Petal_Length", data = irisData, hue = "Species")

**[숙제52]** 
x축은 species y축은 petal_width인 boxplot을 보여준다다

In [ ]:
## Barplot : 숫자 변수의 경향 추청치를 보여준다.

sns.barplot(x="Species", y="Petal_Width", data =irisData)



**[숙제53]** 
x축은 species y축은 sepal_length인 boxplot을 보여준다다

In [ ]:
## Box plot

sns.boxplot(x="Species", y="Sepal_Length", data=irisData)

In [ ]:
 ## sns.boxplot(x="Species", y="Sepal_Length", data=irisData, hue="group") data에 group이 없다.